File: frontend/src/lib/api/tasks.ts
import { z } from 'zod';

// Mirrors the TaskRunAck wire shape from POST /api/tasks/* (snake_case): which task was
// queued and how many units of work (books) it will process. Fire-and-forget — there's
// no live progress, mirroring extraction.
export const taskRunAckSchema = z.object({
	task: z.string(),
	status: z.string(),
	queued: z.number().int().nonnegative()
});

export type TaskRunAck = z.infer<typeof taskRunAckSchema>;

/** Queue AI generation of book-level keywords across the library. `regenerate` re-tags
 *  every extracted book; otherwise only those missing keywords. Fire-and-forget: the
 *  sweep runs on the background worker. `fetchFn` is injectable for SSR/tests. */
export async function triggerBookKeywords(
	regenerate = false,
	fetchFn: typeof fetch = fetch
): Promise<TaskRunAck> {
	const res = await fetchFn('/api/tasks/book-keywords', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ regenerate })
	});
	if (!res.ok) throw new Error(`POST /api/tasks/book-keywords → ${res.status}`);
	return taskRunAckSchema.parse(await res.json());
}

/** Queue an AI-assisted dedup of the whole keyword vocabulary, merging near-duplicate
 *  tags ("Veggie" → "Vegetarian") across recipes and books. Fire-and-forget: the pass
 *  runs on the background worker. `queued` is the vocabulary size it will analyse.
 *  `fetchFn` is injectable for SSR/tests. */
export async function triggerDedupKeywords(fetchFn: typeof fetch = fetch): Promise<TaskRunAck> {
	const res = await fetchFn('/api/tasks/dedup-keywords', { method: 'POST' });
	if (!res.ok) throw new Error(`POST /api/tasks/dedup-keywords → ${res.status}`);
	return taskRunAckSchema.parse(await res.json());
}

/** Queue classification of every keyword that has not yet been examined. */
export async function triggerClassifyKeywords(fetchFn: typeof fetch = fetch): Promise<TaskRunAck> {
	const res = await fetchFn('/api/tasks/classify-keywords', { method: 'POST' });
	if (!res.ok) throw new Error(`POST /api/tasks/classify-keywords → ${res.status}`);
	return taskRunAckSchema.parse(await res.json());
}

/** Queue an AI-assisted merge of canonical ingredient variants. The worker repoints
 * recipe ingredient facts before removing duplicates; `queued` is the vocabulary size. */
export async function triggerDedupIngredients(fetchFn: typeof fetch = fetch): Promise<TaskRunAck> {
	const res = await fetchFn('/api/tasks/dedup-ingredients', { method: 'POST' });
	if (!res.ok) throw new Error(`POST /api/tasks/dedup-ingredients → ${res.status}`);
	return taskRunAckSchema.parse(await res.json());
}

/** Queue an import of new books from the Calibre library into the v2 DB. Existing
 *  books are left unchanged.
 *  Fire-and-forget: the import runs on the background worker and its result lands on
 *  the task run. `queued` is 0 (the book count isn't known until the worker reads the
 *  library). `fetchFn` is injectable for SSR/tests. */
export async function triggerCalibreSync(fetchFn: typeof fetch = fetch): Promise<TaskRunAck> {
	const res = await fetchFn('/api/tasks/calibre-sync', { method: 'POST' });
	if (!res.ok) throw new Error(`POST /api/tasks/calibre-sync → ${res.status}`);
	return taskRunAckSchema.parse(await res.json());
}

File: backend/app/api/tasks.py
from fastapi import APIRouter, status
from sqlalchemy import func, select

from app.db import SessionDep
from app.models.book import Book
from app.models.enums import TaskType
from app.models.ingredient import CanonicalIngredient
from app.models.recipe import Keyword
from app.schemas.tasks import BookKeywordTaskRequest, TaskRunAck
from app.tasks.book_keywords import enqueue_backfill_book_keywords
from app.tasks.calibre_sync import enqueue_calibre_sync
from app.tasks.ingredient_dedup import enqueue_dedup_ingredients
from app.tasks.keyword_classification import enqueue_classify_keywords
from app.tasks.keyword_dedup import enqueue_dedup_keywords
from app.tasks.runs import create_task_run

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.post(
    "/book-keywords",
    response_model=TaskRunAck,
    status_code=status.HTTP_202_ACCEPTED,
)
def trigger_book_keywords(body: BookKeywordTaskRequest, session: SessionDep) -> TaskRunAck:
    """Queue AI generation of book-level keywords across the library. By default tags
    only extracted books that have none yet; `regenerate` re-tags every extracted book.
    Records a queued TaskRun and dispatches to the worker; returns how many books are
    eligible (the count the task will work through), not a live progress handle."""
    stmt = select(func.count()).select_from(Book).where(Book.recipes.any())
    if not body.regenerate:
        stmt = stmt.where(~Book.keywords.any())
    eligible = session.scalar(stmt) or 0

    run = create_task_run(session, TaskType.BOOK_KEYWORDS, detail={"regenerate": body.regenerate})
    enqueue_backfill_book_keywords(str(run.id), body.regenerate)
    return TaskRunAck(task="book_keywords", status="queued", queued=eligible)


@router.post(
    "/dedup-keywords",
    response_model=TaskRunAck,
    status_code=status.HTTP_202_ACCEPTED,
)
def trigger_dedup_keywords(session: SessionDep) -> TaskRunAck:
    """Queue an AI-assisted dedup of the whole keyword vocabulary, merging near-duplicate
    tags ("Veggie" -> "Vegetarian") across recipes and books. Records a queued TaskRun and
    dispatches to the worker; returns the vocabulary size it will analyse."""
    vocabulary = session.scalar(select(func.count()).select_from(Keyword)) or 0
    run = create_task_run(session, TaskType.KEYWORD_DEDUP)
    enqueue_dedup_keywords(str(run.id))
    return TaskRunAck(task="keyword_dedup", status="queued", queued=vocabulary)


@router.post(
    "/classify-keywords",
    response_model=TaskRunAck,
    status_code=status.HTTP_202_ACCEPTED,
)
def trigger_classify_keywords(session: SessionDep) -> TaskRunAck:
    """Queue classification of every keyword not yet examined by the classifier."""
    pending = (
        session.scalar(
            select(func.count()).select_from(Keyword).where(Keyword.classified_at.is_(None))
        )
        or 0
    )
    run = create_task_run(session, TaskType.KEYWORD_CLASSIFICATION)
    enqueue_classify_keywords(str(run.id))
    return TaskRunAck(task="keyword_classification", status="queued", queued=pending)


@router.post(
    "/dedup-ingredients",
    response_model=TaskRunAck,
    status_code=status.HTTP_202_ACCEPTED,
)
def trigger_dedup_ingredients(session: SessionDep) -> TaskRunAck:
    """Queue an AI-assisted merge of near-duplicate canonical ingredients."""
    vocabulary = session.scalar(select(func.count()).select_from(CanonicalIngredient)) or 0
    run = create_task_run(session, TaskType.INGREDIENT_DEDUP)
    enqueue_dedup_ingredients(str(run.id))
    return TaskRunAck(task="ingredient_dedup", status="queued", queued=vocabulary)


@router.post(
    "/calibre-sync",
    response_model=TaskRunAck,
    status_code=status.HTTP_202_ACCEPTED,
)
def trigger_calibre_sync(session: SessionDep) -> TaskRunAck:
    """Queue an import of new books from the Calibre library into the v2 DB. Existing
    books are left unchanged. Records a queued TaskRun and dispatches to the worker;
    the count of books imported isn't known until the worker reads the library, so
    `queued` is 0 — the run's `detail` carries the created/skipped/excluded result
    once it completes."""
    run = create_task_run(session, TaskType.CALIBRE_SYNC)
    enqueue_calibre_sync(str(run.id))
    return TaskRunAck(task="calibre_sync", status="queued", queued=0)
